src/operation/create_indexes/mod.rs

```diff
@@ -0,0 +1,172 @@
+
+#[derive(Debug)]
+pub(crate) struct CreateIndexes {
+    ns: Namespace,
+    indexes: Vec<IndexModel>,
+    options: Option<CreateIndexOptions>,
+}
+
+impl CreateIndexes {
+    pub(crate) fn new(
+        ns: Namespace,
+        indexes: Vec<IndexModel>,
+        options: Option<CreateIndexOptions>,
+    ) -> Self {
+        Self {
+            ns,
+            indexes,
+            options,
+        }
+    }
+
+    #[cfg(test)]
+    pub(crate) fn with_indexes(indexes: Vec<IndexModel>) -> Self {
+        Self {
+            ns: Namespace {
+                db: String::new(),
+                coll: String::new(),
+            },
+            indexes,
+            options: None,
+        }
+    }
+}
+
+impl Operation for CreateIndexes {
+    type O = CreateIndexesResult;
+    type Command = Document;
+    type Response = CommandResponse<Response>;
+    const NAME: &'static str = "createIndexes";
+
+    fn build(&mut self, _description: &StreamDescription) -> Result<Command> {
+        self.indexes.iter_mut().for_each(|i| i.update_name()); // Generate names for unnamed indexes.
+        let indexes = bson::to_bson(&self.indexes)?;
+        let mut body = doc! {
+            Self::NAME: self.ns.coll.clone(),
+            "indexes": indexes,
+        };
+        append_options(&mut body, self.options.as_ref())?;
+
+        Ok(Command::new(
+            Self::NAME.to_string(),
+            self.ns.db.clone(),
+            body,
+        ))
+    }
+
+    fn handle_response(
+        &self,
+        response: Response,
+        _description: &StreamDescription,
+    ) -> Result<Self::O> {
+        let index_names = self.indexes.iter().filter_map(|i| i.get_name()).collect();
+        Ok(CreateIndexesResult {
+            index_names,
+            created_collection_automatically: response.created_collection_automatically,
+            num_indexes_before: response.num_indexes_before,
+            num_indexes_after: response.num_indexes_after,
+            note: response.note,
+            commit_quorum: response.commit_quorum,
+        })
+    }
+
+    fn write_concern(&self) -> Option<&WriteConcern> {
+        self.options
+            .as_ref()
+            .and_then(|opts| opts.write_concern.as_ref())
+    }
+
+    fn selection_criteria(&self) -> Option<&SelectionCriteria> {
+        Some(SelectionCriteria::ReadPreference(ReadPreference::Primary)).as_ref()
+    }
+
+    fn retryability(&self) -> Retryability {
+        Retryability::Read
+    }
+}
+#[derive(Debug)]
+pub(crate) struct Response {
+    created_collection_automatically: Option<bool>,
+    num_indexes_before: u32,
+    num_indexes_after: u32,
+    note: Option<String>,
+    commit_quorum: Option<CommitQuorum>,
+}
+
+impl<'de> Deserialize<'de> for Response {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        struct ResponseBody {
```

Comment:
`ResponseBody` and `Response` look to be identical - is there a need for the former?

Reply:
It's necessary because the response after running `createIndexes` depends on the toplogy.
For example, running `createIndexes` on a single toplogy might yield
```
{ numIndexesBefore: 1, numIndexesAfter: 2 }
```
but running on a sharded topology could yield 
```
{ raw: { mongodb://localhost:27017/: { numIndexesBefore: 1, numIndexesAfter: 2 } } }
```
Because the internals of the response are the same, we have to do a custom deserialization in order to account for the wrapper around the "relevant" part of the response.

Reply:
Oh, I see, thank you - I'd missed the need for two different implementations of `Deserialize`.  Could also achieve that with a newtype, but that's not a must.